### DISCUSSION TITLE
Disk used for file uploaded is hardcoded in testing environments

### DISCUSSION THREAD

**Discussion by kajgies** [livewire/livewire#3105] [7. Bugs (archived)]
The disk used in testing environments is currently hardcoded with no override possible:
FileUploadConfiguration.php Line 27:
```
    public static function disk()
    {
        if (app()->environment('testing')) {
            return 'tmp-for-tests';
        }

        return config('livewire.temporary_file_upload.disk') ?: config('filesystems.default');
    }
```
This means that in a testing environment you have no choice to add a 'tmp-for-tests' disk to your filesystems config, otherwise file uploads will give an error.

I think it would be a good idea to remove this hardcoded requirement and make the disk environment specific in the config using env(),

**Comment by kajgies** [comment]
Also the cause of this issue:  #1341

**Comment by PhiloNL** [comment]
@kajgies could you leave a comment at https://github.com/livewire/livewire/pull/2571#issuecomment-788035594 with some more information why this is something you would like to see? 😄

**Comment by ohnotnow** [comment]
I think we're hitting an issue with this while testing and using a combination of S3 and the Spatie Media Library.  In the code we have a call to :
```php
$booking->addMediaFromDisk($this->file->path(), config('website.default_disk', 'images'))
    ->toMediaCollection('booking-image');
```
And in the tests we do a `Storage::fake()` call and set the various `config()` entries to all match.  

But in the test it fails as the `$this->file->path()` call will return something like 
```
/Users/ohwhatnow/Documents/code/website/storage/framework/testing/disks/tmp-for-tests/livewire-tmp/2dGiPtHhSx9JjzOXQFWstUFVQZl6XB-metacGdqb2IucG5n--size=204.8.png
```

But in a production env it would return a relative-to-the-disk path() like 
```
livewire-tmp/WrU1Tzh7fbgexvLkegMqV452C0fHlV-metaMTAyNDczODlfNjMyNDM3OTIwMTQzODgwXzcwODgyNzczODA5MTQzNTg1MDdfbi5qcGc=-.jpg
```  
So when the Spatie package checks for the file existing in the configured disk it fails.

**Comment by joshhanley** [comment]
@ohnotnow this PR #4001 was merged in and tagged in 2.7.0, which I think should fix that issue with Media Library.

Can you test it and confirm whether it works for you or not?

Hope this helps!

**Comment by ohnotnow** [comment]
@joshhanley I'm afraid not - livewire still seems to use it's own disk during the test so I still get (on livewire 2.7.2) :
```
dd($this->file->path());
"/Users/ohwhatnow/Documents/code/website/storage/framework/testing/disks/tmp-for-tests/livewire-tmp/blah.png"
```
That's after setting my own config values and `Storage::fake('images')` in the test and I've also tried manually setting 
```
config(['livewire.temporary_file_upload.disk' => 'images']);
```
It still looks to be hard-coded in `FileUploadConfiguration.php` to use livewire's own disk :
```php
    public static function disk()
    {
        if (app()->runningUnitTests()) {
            return 'tmp-for-tests';
        }

        return config('livewire.temporary_file_upload.disk') ?: config('filesystems.default');
    }
```

**Reply by joshhanley** [reply]
@ohnotnow oh right, sorry I misread the bit about in tests 🤦‍♂️ yeah that is a different issue.

What is the issue using the `tmp-for-tests` disk? Also can you give an example of a test and what you expect to happen?

Thanks!

**Comment by ohnotnow** [comment]
Hi @joshhanley - the issue is that my code uses the spatie media library and should upload the temp files to an S3 bucket, then get moved to another once it's done.  In the app we have a `config('theapp.default_disk')` which is where we tell the media library code where to look (copied from above) :
```php
$booking->addMediaFromDisk($this->file->path(), config('website.default_disk', 'images'))
    ->toMediaCollection('booking-image');
```
and in the test we just do a plain `Livewire::test()...->set('file', $fakeUploadedFile)->call('save');` - but because the file has been uploaded to livewire's own `tmp-for-tests` disk the call to the `->addMediaFromDisk()` fails as the path to the file isn't based on the right disk (it returns the full `/absolute/path/to/local/file` rather than an S3-disk which returns `relative/path/to/file`).

In that specific test I _might_ be able to set our `default_disk` to `tmp-for-tests` but it feels a bit icky - and I know I have some more 'integration' tests where it's going to get even messier.

Is there a reason livewire does the `tmp-for-tests` thing?

**Reply by joshhanley** [reply]
@ohnotnow yep makes sense.

Can you give me an example of a complete test (test code only)?

As to why it defaults to `tmp-for-tests`, I'm not sure 🙂 possibly ease of use for local file uploads maybe?

Thanks!

**Comment by ohnotnow** [comment]
@joshhanley no problem - I've shoved a little repo up at https://github.com/ohnotnow/livewire-tmp-disk with `tests/Feature/FileFormTest.php` in it and a stripped down `app/Http/Livewire/FileForm.php` component that should demonstrate (I hope!) what I'm trying to do and the error I'm getting when using a non-tmp disk.

**Reply by joshhanley** [reply]
@ohnotnow thanks! Finally got a chance to pull this down and have a play with out. Removing the conditional and `tmp-for-tests` in the below code makes your test pass.

https://github.com/livewire/livewire/blob/266a72d0b6601af0982f62e9bdeca5bcff2e1048/src/FileUploadConfiguration.php#L29-L31

The main issue is because the `addMediaFromDisk` method below accepts the disk name, `s3` for the temp directory. This is then  different to where the tests are storing temp files, which is storing them in `tmp-for-tests` directory.

```php
User::first()->addMediaFromDisk($this->file->path(), 's3')->toMediaCollection();
```

The problem is, if we make that change now and remove it, it will be a breaking change, as people who haven't faked their temp disk will start seeing it use their real temp directory (which could be S3).

I'm not sure how we can work around that yet, so open to ideas if you have any.

**Reply by joshhanley** [reply]
@ohnotnow I think I might have a potential solution, what if instead of `return 'tmp-for-tests';` we just do this, in `FileUploadConfiguration.php`

```php
if (app()->runningUnitTests()) {
    config()->set('livewire.temporary_file_upload.disk', 'tmp-for-tests');
}
```

Then in your app, rather than hardcoding `'s3'`, you use the config value, like:

```php
User::first()->addMediaFromDisk($this->file->getRelativePath(), config('livewire.temporary_file_upload.disk'))->toMediaCollection();
```

The other piece that would be needed is a new helper `->getRelativePath()`, which you can see used above.

This would be rather than `->path()` which returns an absolute path for local disks and relative path for s3 disks. See Laravel docs for details on that https://laravel.com/docs/8.x/filesystem#file-paths

> You may use the path method to get the path for a given file. If you are using the `local` driver, this will return the absolute path to the file. If you are using the s3 driver, this method will return the `relative` path to the file in the S3 bucket

What do you think?

**Reply by ohnotnow** [reply]
@joshhanley sorry for not replying! GH didn't give me any notification about your replies!  I was just looking at this again as I was going to do a PR myself :-)

I was thinking of doing something like
```php
if (app()->runningUnitTests() && ! config('livewire.temporary_file_upload.disk')) {
    config()->set('livewire.temporary_file_upload.disk', 'tmp-for-tests');
}
```
I realise - as you've said - that if someone hasn't faked their disk then it'd cause them problems though.  Hrm.  Maybe this should be a 'v3' thing?  We want to be able to test that it does upload to the disk we specify as otherwise other bits of code/tests would need to be aware of livewire tmp-for-tests too rather than our global `app.default_disk`.

For instance - if that `->addMediaFromDisk()` call is in a generic queued 'process uploaded image' job rather than directly in the component - it would now need to know 'mibbe this was a livewire upload, mibbe not'.  Or we could start passing the disk name as a parameter I guess, but I always get a little antsy changing code up to work around an issue with tests.

Really not sure other than having the breaking change.  I'd _guess_ the number of people using S3 for uploads and _not_ faking the disks would be a fairly small % of people.  At the same time, having them find out they had forgotten to fake something and had now run up a $1000 bill on S3 during some CI runs would be ... less than ideal.

**Comment by edalzell** [comment]
I'm now hitting this, same scenario. I fake my `s3` storage so it doesn't save the file, but getting the same `tmp-for-test` not defined error.

**Reply by yilanboy** [reply]
Same here!

**Comment by boptom** [comment]
Getting this now when not specifying a disk when saving a file. i.e. using the disk set in config.

Also, seems to break tests when deleting a file.

**Comment by CamKem** [comment]
Has this issue become a problem again? I have noticed in the logs, when working on unit tests an image upload feature and any unit test where I set the value of a property to a fake UploadedFiled is triggering an error in the logs for: 

[2024-07-18 08:06:47] testing.ERROR: Disk [tmp-for-tests] does not have a configured driver. {"userId":1,"exception":"[object] (InvalidArgumentException(code: 0): Disk [tmp-for-tests] does not have a configured driver. at /Users/cam/Sites/pinkary-repo/vendor/laravel/framework/src/Illuminate/Filesystem/FilesystemManager.php:137)

This is an example test & I can provide the full stack trace if needed.

```php
test('used images are NOT cleanup when store is called', function () {
    Storage::fake('public');
    $user = User::factory()->create();
    $file = UploadedFile::fake()->image('photo1.jpg');
    $name = $file->hashName();
    $date = now()->format('Y-m-d');
    $path = 'images/'.$date.'/'.$name;

    $component = Livewire::actingAs($user)->test(Create::class, [
        'toId' => $user->id,
    ]);
    $component->set('images', [$file]);

    Storage::disk('public')->assertExists($path);

    expect(session('images'))->toBeArray()
        ->and(session('images'))->toContain($path);

    $url = Storage::disk('public')->url($path);

    $component->set('content', "![Image Alt Text]({$url})");
    $component->call('store');

    Storage::disk('public')->assertExists($path);
    expect(session('images'))->toBeNull();
});
```
